### PR TITLE
Modify fee of System.Runtime.GetNotifications

### LINKED ISSUE #2725: GAS price for GetNotifications is cheap
The syscall `System.Runtime.GetNotifications` can get all notifications of a contract with a limit of 2048.
Although the limit is 2048, lots of small objects' allocation and GC could consume some time.

This attack can be prevented by @Liaojinghui's method if we do the same thing to syscall as what we do to opcode in https://github.com/neo-project/neo/pull/2724.

POC: `VwEAwgwFYWFhYWEB9AF3AEtLQZUBb2FvAJ13AG8AJfL///9JQdv+qHQCAJQ1d3cASkEnQzXxRW8AnXcAbwAl8v///0lA`

43 minutes cost in this case, source code can be seen in [Tanya](https://github.com/lazynode/Tanya/pull/30/files)

Essentially, using an under-priced opcode or syscall once and once again should be blamed.

### PR DESCRIPTION
**Pull request by erikzhang**
1. Fixed #2725
2. Added a benchmark.

### CODE REVIEW

**Comment by Liaojinghui** [comment]
Really happy to see we finally have a benchmark project.

**Comment by erikzhang** [comment]
> it would be better if we could use [BenchmarkDotNet](https://benchmarkdotnet.org/Overview.htm)

If the project become more complex in the future, we can consider it. Currently I don't think it's needed much.

File: src/neo/SmartContract/ApplicationEngine.Runtime.cs [hunk 1]
```diff
@@ -122,7 +122,7 @@ partial class ApplicationEngine
         /// The <see cref="InteropDescriptor"/> of System.Runtime.GetNotifications.
         /// Gets the notifications sent by the specified contract during the execution.
         /// </summary>
-        public static readonly InteropDescriptor System_Runtime_GetNotifications = Register("System.Runtime.GetNotifications", nameof(GetNotifications), 1 << 8, CallFlags.None);
+        public static readonly InteropDescriptor System_Runtime_GetNotifications = Register("System.Runtime.GetNotifications", nameof(GetNotifications), 1 << 12, CallFlags.None);
```
Comment (roman-khimov) [inline src/neo/SmartContract/ApplicationEngine.Runtime.cs:125]: Isn't it a hardfork?

Comment (shargon) [reply src/neo/SmartContract/ApplicationEngine.Runtime.cs:125]: Yes, i think so

**Comment by roman-khimov** [comment]
Maybe we should charge per-element here? Sending notifications is naturally charged per-element, maybe getting them should also be, it'd be more fair.

**Comment by erikzhang** [comment]
The problem with per-element charges is that when you create a transaction it is impossible to know how many notifications the contract will receive when it executes.

**Comment by roman-khimov** [comment]
Transaction scripts are test-executed usually and I don't think this problem is any different from any other test/real invocation GAS/state difference that can happen in lots of other scenarios.

**Comment by erikzhang** [comment]
> test/real invocation GAS/state difference

We should try to avoid this situation.

**Comment by roman-khimov** [comment]
> > test/real invocation GAS/state difference
> 
> We should try to avoid this situation.

Sure, but here it will depend on the number of emitted notifications. To me it's not very likely to change unless the expected transaction outcome changes significantly anyway (like some transfer or asset mint fails). Even if it is to change, it's more likely that some notifications will be omitted rather than created.

**Comment by erikzhang** [comment]
@roman-khimov What about neo-project/neo-vm#473?

**Comment by roman-khimov** [comment]
> @roman-khimov What about [neo-project/neo-vm#473](https://github.com/neo-project/neo-vm/pull/473)?

Unexpected. What's the execution time for #2725 with this? In some ways it feels like this problem shouldn't require changing the VM, but at the same time it can be effective and not very restrictive (it's hard to imagine normal contract trying to reuse (referenced) event items after notification).

**Comment by erikzhang** [comment]
Immutability can be very useful, not only in the notification's scenario, but also for calling between contracts. We can open this function to contracts in the future, such as adding a new `OpCode.ASIMMUTABLE`.

File: src/neo/SmartContract/ApplicationEngine.Runtime.cs (hunk 1 above)
Comment (superboyiii) [reply src/neo/SmartContract/ApplicationEngine.Runtime.cs:125]: Full sync mainnet, storage doesn't fork.

**Comment by erikzhang** [comment]
Merge?

**Review by shargon — APPROVED** [review]

**Review by Liaojinghui — APPROVED** [review]

**Pull request merged**